LanguageExt.Core/DataTypes/Seq/Enum.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace LanguageExt
+{
+    internal class Enum<A>
+    {
+        IEnumerator<A> iter;
+        List<A> list;
+
+        public Enum(IEnumerable<A> seq)
+        {
+            this.iter = seq.GetEnumerator();
+            this.list = new List<A>();
+        }
+
+        public Enum(IEnumerator<A> iter)
+        {
+            this.iter = iter;
+            this.list = new List<A>();
+        }
+
+        ~Enum()
+        {
+            iter?.Dispose();
+            iter = null;
+        }
+
+        public (bool Success, A Value) Get(int index)
+        {
+            if (index < list.Count)
+            {
+                return (true, list[index]);
+            }
+            if (iter == null)
+            {
+                return (false, default(A));
+            }
+            else
+            {
+                lock (list)
+                {
+                    bool theresMore = true;
+                    while (index >= list.Count && (theresMore = iter.MoveNext()))
```

Comment:
Do we want to move the `lock(list)` down here so we unlock between each iteration of `.Add` in case the incoming iterator is really long and you've asked for a really high `index`?

Reply:
I guess there's one of two trade-offs needs to be picked here:

1. The lazy stream blocks if you try to jump ahead 
2. The general performance if lazy streams is reduced from the cost of the `lock` acquire per iteration

Going to have a think on this one.  And probably do some perf tests to see where we stand.  I'm erring towards your solution, because the type will act more like a regular immutable type.

Reply:
I have made the change you suggested.